File: iCatchUp/iCatchUp/Network/NewsApi.swift
//
//  NewsApi.swift
//  iCatchUp
//
//  Created by Developer User on 10/18/18.
//  Copyright © 2018 UPC. All rights reserved.
//

import Foundation
import os
import Alamofire



class NewsApi {
    static let baseUrl = "https://newsapi.org"
    static let topHeadlinesUrl = "\(baseUrl)/v2/top-headlines"
    static let everythingUrl = "\(baseUrl)/v2/everything"
    static let sourcesUrl = "\(baseUrl)/v2/sources"
    
    static private func get<T: Decodable>(
        from urlString: String,
        parameters: [String : String],
        responseType: T.Type,
        responseHandler: @escaping ((T) -> Void),
        errorHandler: @escaping ((Error) -> Void)) {
        // Validate URL
        guard let url = URL(string: urlString) else {
            let message = "Error on URL"
            os_log("%@", message)
            return
        }
        // Make the Request
        Alamofire.request(url, parameters: parameters)
            .validate()
            .responseJSON(completionHandler: { response in
            switch response.result {
            case .success(_):
                do {
                    let decoder = JSONDecoder()
                    if let data = response.data {
                        let dataResponse = try decoder.decode(responseType, from: data)
                        responseHandler(dataResponse)
                    }
                } catch {
                    errorHandler(error)
                }
            case .failure(let error):
                errorHandler(error)
            }
        })
    }
    
    static func getSources(
        apiKey: String,
        responseHandler: @escaping ((SourcesResponse) -> Void),
        errorHandler: @escaping ((Error) -> Void)) {
        let parameters = ["apiKey" : apiKey]
        self.get(
            from: sourcesUrl,
            parameters: parameters,
            responseType: SourcesResponse.self,
            responseHandler: responseHandler,
            errorHandler: errorHandler)
    }
    
    static func getTopHeadlines(
        apiKey: String,
        responseHandler: @escaping ((ArticlesResponse) -> Void),
        errorHandler: @escaping ((Error) -> Void)) {
        let parameters = ["apiKey" : apiKey, "country" : "us"]
        self.get(
            from: topHeadlinesUrl,
            parameters: parameters,
            responseType: ArticlesResponse.self,
            responseHandler: responseHandler,
            errorHandler: errorHandler)
    }
    
    static func getEverything(
        apiKey: String,
        fromSources sources: String?,
        responseHandler: @escaping ((ArticlesResponse) -> Void),
        errorHandler: @escaping ((Error) -> Void)) {
        var parameters = ["apiKey" : apiKey]
        if let sources = sources {
            parameters["sources"] = sources
        }
        self.get(
            from: everythingUrl,
            parameters: parameters,
            responseType: ArticlesResponse.self,
            responseHandler: responseHandler,
            errorHandler: errorHandler)
    }
}

File: CatchUp/app/src/main/java/pe/edu/upc/catchup/network/NewsApi.kt
package pe.edu.upc.catchup.network

import com.androidnetworking.AndroidNetworking
import com.androidnetworking.common.Priority
import com.androidnetworking.error.ANError
import com.androidnetworking.interfaces.ParsedRequestListener
import pe.edu.upc.catchup.R

class NewsApi {
    companion object {
        private val baseUrl = "https://newsapi.org"
        val topHeadlinesUrl = "$baseUrl/v2/top-headlines"
        val everythingUrl = "$baseUrl/v2/everything"
        val sourcesUrl = "$baseUrl/v2/sources"
        val tag = "CatchUp"

        fun requestHeadlines(key: String, responseHandler: (ArticlesResponse?) -> Unit, errorHandler: (ANError?) -> Unit  ) {
            AndroidNetworking.get(NewsApi.topHeadlinesUrl)
                    .addQueryParameter("apiKey", key)
                    .addQueryParameter("country", "us")
                    .setPriority(Priority.LOW)
                    .setTag(tag)
                    .build()
                    .getAsObject(ArticlesResponse::class.java, object : ParsedRequestListener<ArticlesResponse> {
                        override fun onResponse(response: ArticlesResponse?) {
                            responseHandler(response)
                        }

                        override fun onError(anError: ANError?) {
                            errorHandler(anError)
                        }

                    })
        }
        fun requestEverythingFromSources(key: String, sourceIds: String, responseHandler: (ArticlesResponse?) -> Unit, errorHandler: (ANError?) -> Unit  ) {
            AndroidNetworking.get(NewsApi.everythingUrl)
                    .addQueryParameter("apiKey", key)
                    .addQueryParameter("sources", sourceIds)
                    .setPriority(Priority.LOW)
                    .setTag(tag)
                    .build()
                    .getAsObject(ArticlesResponse::class.java, object : ParsedRequestListener<ArticlesResponse> {
                        override fun onResponse(response: ArticlesResponse?) {
                            responseHandler(response)
                        }

                        override fun onError(anError: ANError?) {
                            errorHandler(anError)
                        }

                    })
        }
        fun requestSources(key: String, responseHandler: (SourcesResponse?) -> Unit, errorHandler: (ANError?) -> Unit  ) {
            AndroidNetworking.get(NewsApi.sourcesUrl)
                    .addQueryParameter("apiKey", key)
                    .setPriority(Priority.LOW)
                    .setTag(tag)
                    .build()
                    .getAsObject(SourcesResponse::class.java, object : ParsedRequestListener<SourcesResponse> {
                        override fun onResponse(response: SourcesResponse?) {
                            responseHandler(response)
                        }

                        override fun onError(anError: ANError?) {
                            errorHandler(anError)
                        }
                    })

        }
    }

}